File: 练习/2025-12/统计被覆盖的建筑/统计被覆盖的建筑.py
from collections import defaultdict
from typing import List

class Solution:
    def countCoveredBuildings(self, n: int, buildings: List[List[int]]) -> int:
        ans = 0
        #预处理每行最大最小 每列最大最小
        row_min = defaultdict(lambda: int(1e9)) 
        row_max = defaultdict(lambda: -1)
        col_min = defaultdict(lambda: int(1e9))
        col_max = defaultdict(lambda: -1)

        for x,y in buildings:
            row_min[x] = min(row_min[x], y)
            row_max[x] = max(row_max[x], y)
            col_min[y] = min(col_min[y], x)
            col_max[y] = max(col_max[y], x)

        for x,y in buildings:
            if row_min[x] < y < row_max[x] and col_min[y] < x < col_max[y]:
                ans += 1

        return ans

File: 练习/2025-12/统计被覆盖的建筑/统计被覆盖的建筑.java
class Solution {
    public int countCoveredBuildings(int n, int[][] buildings) {
        int ans = 0;
        Map<Integer,Integer> row_min = new HashMap<>();
        Map<Integer,Integer> row_max = new HashMap<>();
        Map<Integer,Integer> col_min = new HashMap<>();
        Map<Integer,Integer> col_max = new HashMap<>();

        for(int[] building : buildings){
            int x = building[0];
            int y = building[1];
            row_min.put(x, Math.min(row_min.getOrDefault(x, Integer.MAX_VALUE), y));
            row_max.put(x, Math.max(row_max.getOrDefault(x, Integer.MIN_VALUE), y));
            col_min.put(y, Math.min(col_min.getOrDefault(y, Integer.MAX_VALUE), x));
            col_max.put(y, Math.max(col_max.getOrDefault(y, Integer.MIN_VALUE), x));
        }

        for(int[] building : buildings)
        {
            int x = building[0];
            int y = building[1];
            boolean f1 = (x > col_min.get(y) && x < col_max.get(y));
            boolean f2 = (y > row_min.get(x) && y < row_max.get(x));

            if(f1 && f2)
            {
                ans++;
            }
        }
        return ans;
    }
}